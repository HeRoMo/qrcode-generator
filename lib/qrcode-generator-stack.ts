import { Stack, StackProps } from 'aws-cdk-lib';
import {
  NodejsFunction as LambdaFunc,
} from 'aws-cdk-lib/aws-lambda-nodejs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { LambdaRestApi } from 'aws-cdk-lib/aws-apigateway';
import { Construct } from 'constructs';

export class QrcodeGeneratorStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const runtime = Runtime.NODEJS_16_X;

    // Lambda Function
    const func = new LambdaFunc(this, 'QRCodeGenerator', {
      runtime,
      entry: 'src/index.ts',
      handler: 'handler',
      bundling: {
        minify: true,
        sourceMap: true,
        target: 'es2020',
      },
    });

    // APIGateway
    const api = new LambdaRestApi(this, 'QRCodeGeneratorAPI', {
      handler: func,
      restApiName: 'qrcode',
      proxy: false,
    });
    api.root.addMethod('GET');
  }
}
