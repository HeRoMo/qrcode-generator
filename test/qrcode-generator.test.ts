import { Template } from '@aws-cdk/assertions';
import { App, Stack } from '@aws-cdk/core';
import { QrcodeGeneratorStack } from '@lib/qrcode-generator-stack';

describe('Snapshot Test', () => {
  it('match snapshot', () => {
    const app = new App();
    const stack = new QrcodeGeneratorStack(app, 'MyTestStack');
    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});

describe('Contain specific resources', () => {
  let template: Template;
  beforeEach(() => {
    const app = new App();
    const stack = new QrcodeGeneratorStack(app, 'MyTestStack');
    template = Template.fromStack(stack);
  });
  it('have Lambda Function', () => {
    template.hasResourceProperties('AWS::Lambda::Function', {
      Runtime: 'nodejs14.x',
    });
  });
  it('have ApiGateway', () => {
    template.hasResourceProperties('AWS::ApiGateway::RestApi', {
      Name: 'qrcode',
    });
  });
});
