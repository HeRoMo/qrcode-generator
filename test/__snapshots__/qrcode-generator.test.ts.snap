// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test match snapshot 1`] = `
Object {
  "Outputs": Object {
    "QRCodeGeneratorAPIEndpoint187B9CEC": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "QRCodeGeneratorAPI2C9FA9D6",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "QRCodeGeneratorAPIDeploymentStageprod9D06B8BC",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters43fd7c63b6d6ab04bac4c556557d54fa8ff22b6b35700c189089ed02b8ae633aArtifactHash5A2044C1": Object {
      "Description": "Artifact hash for asset \\"43fd7c63b6d6ab04bac4c556557d54fa8ff22b6b35700c189089ed02b8ae633a\\"",
      "Type": "String",
    },
    "AssetParameters43fd7c63b6d6ab04bac4c556557d54fa8ff22b6b35700c189089ed02b8ae633aS3Bucket97F1A301": Object {
      "Description": "S3 bucket for asset \\"43fd7c63b6d6ab04bac4c556557d54fa8ff22b6b35700c189089ed02b8ae633a\\"",
      "Type": "String",
    },
    "AssetParameters43fd7c63b6d6ab04bac4c556557d54fa8ff22b6b35700c189089ed02b8ae633aS3VersionKey5537ED42": Object {
      "Description": "S3 key for asset version \\"43fd7c63b6d6ab04bac4c556557d54fa8ff22b6b35700c189089ed02b8ae633a\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "QRCodeGeneratorAPI2C9FA9D6": Object {
      "Properties": Object {
        "Name": "qrcode",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "QRCodeGeneratorAPIAccount8AC74895": Object {
      "DependsOn": Array [
        "QRCodeGeneratorAPI2C9FA9D6",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "QRCodeGeneratorAPICloudWatchRole4704F540",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "QRCodeGeneratorAPICloudWatchRole4704F540": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "QRCodeGeneratorAPIDeploymentA16CD618a99b38e0ec99fd9178d0a360371ed814": Object {
      "DependsOn": Array [
        "QRCodeGeneratorAPIGET656555B1",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "QRCodeGeneratorAPI2C9FA9D6",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "QRCodeGeneratorAPIDeploymentStageprod9D06B8BC": Object {
      "DependsOn": Array [
        "QRCodeGeneratorAPIAccount8AC74895",
      ],
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "QRCodeGeneratorAPIDeploymentA16CD618a99b38e0ec99fd9178d0a360371ed814",
        },
        "RestApiId": Object {
          "Ref": "QRCodeGeneratorAPI2C9FA9D6",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "QRCodeGeneratorAPIGET656555B1": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "QRCodeGeneratorF4F29900",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "QRCodeGeneratorAPI2C9FA9D6",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "QRCodeGeneratorAPI2C9FA9D6",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "QRCodeGeneratorAPIGETApiPermissionMyTestStackQRCodeGeneratorAPICE74E176GETB95CA280": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "QRCodeGeneratorF4F29900",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "QRCodeGeneratorAPI2C9FA9D6",
              },
              "/",
              Object {
                "Ref": "QRCodeGeneratorAPIDeploymentStageprod9D06B8BC",
              },
              "/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "QRCodeGeneratorAPIGETApiPermissionTestMyTestStackQRCodeGeneratorAPICE74E176GETAEA96365": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "QRCodeGeneratorF4F29900",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "QRCodeGeneratorAPI2C9FA9D6",
              },
              "/test-invoke-stage/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "QRCodeGeneratorF4F29900": Object {
      "DependsOn": Array [
        "QRCodeGeneratorServiceRole2960A75B",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters43fd7c63b6d6ab04bac4c556557d54fa8ff22b6b35700c189089ed02b8ae633aS3Bucket97F1A301",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters43fd7c63b6d6ab04bac4c556557d54fa8ff22b6b35700c189089ed02b8ae633aS3VersionKey5537ED42",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters43fd7c63b6d6ab04bac4c556557d54fa8ff22b6b35700c189089ed02b8ae633aS3VersionKey5537ED42",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "QRCodeGeneratorServiceRole2960A75B",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "QRCodeGeneratorServiceRole2960A75B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
